### PR TITLE
splits payer contract similar to project payer

### PR DESCRIPTION
**Pull request by mejango**
add a direct payment contract that splits between JBSplits

### CODE REVIEW

File: contracts/JBETHERC20SplitsPayer.sol [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            _split
+          );
+          // Trigger the allocator's `allocate` function.
+          _split.allocator.allocate{value: _settleAmount}(_data);
+          // Otherwise, if a project is specified, make a payment to it.
+        } else if (_split.projectId != 0) {
+          _pay(
+            _split.projectId,
+            _token,
+            _amount,
```
Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:246]: _settleAmount instead?

File: contracts/JBETHERC20SplitsPayer.sol [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+          // Create the data to send to the allocator.
+          JBSplitAllocationData memory _data = JBSplitAllocationData(
+            _settleAmount,
+            _decimals,
+            _PROTOCOL_PROJECT_ID,
+            0,
+            _split
+          );
+          // Trigger the allocator's `allocate` function.
+          _split.allocator.allocate{value: _settleAmount}(_data);
```
Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:240]: ``
```suggestion
          _split.allocator.allocate{value: _token == JBTokens.ETH ? _settleAmount : 0}(_data);
```

File: contracts/JBETHERC20SplitsPayer.sol (hunk 1 above)
Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:246]:
```suggestion
            _settleAmount,
```

File: contracts/JBETHERC20SplitsPayer.sol (hunk 2 above)
Comment (drgorillamd) [reply contracts/JBETHERC20SplitsPayer.sol:240]: meh, stack too deep then ':)

Comment (drgorillamd) [reply contracts/JBETHERC20SplitsPayer.sol:240]: Suggest then
```
          // Trigger the allocator's `allocate` function (avoiding stack too deep)
          if(_token == JBTokens.ETH) _split.allocator.allocate{value: _settleAmount }(_data);
          else _split.allocator.allocate(_data);
  ```

File: contracts/JBETHERC20SplitsPayer.sol [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+    } else {
+      _payer = address(this);
+      _amount = msg.value;
+      _decimals = 18;
+    }
+
+    // Route the payment to the splits.
+    // Pay the split and get a reference to the amount paid.
+    uint256 _leftoverAmount = _payToSplits(
+      defaultProjectId,
```
Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:224]: `defaultSplitsProjectId` instead? (defaultProjectId is inherited from project payer)

File: contracts/JBETHERC20SplitsPayer.sol [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+
+      if (_splitAmount > 0) {
+        // Transfer tokens to the split.
+        // If there's an allocator set, transfer to its `allocate` function.
+        if (_split.allocator != IJBSplitAllocator(address(0))) {
+          // Create the data to send to the allocator.
+          JBSplitAllocationData memory _data = JBSplitAllocationData(
+            _payer,
+            _splitAmount,
+            0,
```
Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:366]: Shouldn't we pass the decimal? One allocator can have multiple splitsPayer/multiple token, no?

File: contracts/JBETHERC20SplitsPayer.sol [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+            defaultProjectId,
```
Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:367]: same, `defaultSplitsProjectId` instead?

Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:367]: `_splitsProjectId` ?

File: contracts/JBETHERC20SplitsPayer.sol (hunk 4 above)
Comment (mejango) [reply contracts/JBETHERC20SplitsPayer.sol:366]: we should def pass the decimal.

File: contracts/JBETHERC20SplitsPayer.sol (hunk 3 above)
Comment (mejango) [reply contracts/JBETHERC20SplitsPayer.sol:224]: Leave it as is. The `defaultSplitsProjectId` is just used to index the project to which the splits belong. This should, in most practical cases I can think of, be the same as the `defaultProjectId` inherited. But deployers may wanna customize the payout of leftover some other way.

File: contracts/JBETHERC20SplitsPayer.sol [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+          18, // decimals.
+          defaultBeneficiary,
+          0, // min returned tokens.
+          defaultPreferClaimedTokens,
+          defaultMemo,
+          defaultMetadata
+        );
+    // If no project was specified, send the funds directly to the beneficiary or the msg.sender.
+    else
+      Address.sendValue(
```
Comment (drgorillamd) [inline contracts/JBETHERC20SplitsPayer.sol:153]: sendValue will only work to send eth

Comment (mejango) [reply contracts/JBETHERC20SplitsPayer.sol:153]: nice. push going up

**Review by drgorillamd — APPROVED** [review]

**Pull request merged**